Issue8758 migrate nRF sensor temp config to dts

gamnes opened the pull request:
1. Added Temp device node to nRF5 devices with yaml binding.
2. Removed Kconfig Temp priority setting, modified temp_nrf5.c driver to use DT defines.
3. Modified temp_nrf5.c driver to use DT base address define.
4. Added new nrf52 board temperature sensor sample.

Related to #8758.

gamnes commented on drivers/sensor/nrf5/temp_nrf5.c:
```diff
@@ -82,11 +84,10 @@ static int temp_nrf5_channel_get(struct device *dev,
 	struct temp_nrf5_data *data = dev->driver_data;
 
-	temp->EVENTS_DATARDY = 0;
+	nrf_temp_addr->EVENTS_DATARDY = 0;
```
Not using the nRFx HAL nrf_temp.h since this looks a bit outdated. Let me know if this is something we should be doing.

gamnes commented on drivers/sensor/nrf5/temp_nrf5.c:
```diff
@@ -95,18 +96,10 @@ static const struct sensor_driver_api temp_nrf5_driver_api = {
-		    &temp_nrf5_driver, NULL,
-		    POST_KERNEL,
-		    CONFIG_SENSOR_INIT_PRIORITY, &temp_nrf5_driver_api);
+DEVICE_DECLARE(temp_nrf5);
```
Introduced DEVICE_DECLARE, and moved DEVICE_AND_API_INIT() to the bottom to match other implementations (such as UARTE, and HW PWM drivers)

gamnes commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -133,6 +134,18 @@
+		 * .. Default enabled,
+		 * .. Interrupt priority:
+		 * .. .. Range: 0-1, Default: 1
+		 */
```
Added these comments in the DTSI file; not sure if this is the practice that we want to have, or if this comment (interrupt range) should go somewhere else? Can the YAML description enforce this restriction?

gamnes replied:
True - i'll remove it.

dbkinder commented on samples/boards/nrf52/sensor_temperature/README.rst:
```diff
@@ -0,0 +1,35 @@
+   :goals: build flash
+   :compact:
+
+After flashing the image to the board, the user LED on the board display the
```
```suggestion
After flashing the image to the board, the user LEDs on the board display the
```

gamnes replied:
(Y)

dbkinder commented on samples/boards/nrf52/sensor_temperature/README.rst:
```diff
@@ -0,0 +1,35 @@
+The Temperature example shows how to enable and use the embedded nRF5
+TEMP peripheral to read out the DIE temperature of the device. The example
+also enables and uses the LEDs to demonstrate the temperature reading
+passing certain limits (5,20,25,30). The example also outputs the
```
```suggestion
passing certain (Celsius) limits: 5, 20, 25, and 30 degrees. The example also outputs the
```

gamnes replied:
(Y) rewritten

dbkinder commented on samples/boards/nrf52/sensor_temperature/README.rst:
```diff
@@ -0,0 +1,35 @@
+   :compact:
+
+After flashing the image to the board, the user LED on the board display the
+current temperature range (> 5, > 20, > 25, > 30). The board will
```
```suggestion
current temperature range (in |deg| C): >= 5, >= 20, >= 25, and >= 30. The board will
```

gamnes replied:
changed

dbkinder commented on samples/boards/nrf52/sensor_temperature/src/main.c:
```diff
@@ -0,0 +1,93 @@
+			printk("ERROR: Sensor value read failed. \n");
+			return;
+		}
+		printk("sensor value: %d, %d.\n",
```
Maybe mention the temp is in degrees Celsius:
```suggestion
		printk("sensor value: %d C, %d C.\n",
```

gamnes replied:
(Y)

dbkinder reviewed (approved):
LGTM, thanks!

anangl commented on samples/boards/nrf52/sensor_temperature/src/main.c:
```diff
@@ -0,0 +1,97 @@
+/* 1000 msec = 1 sec */
+#define SLEEP_TIME	1000
+
+#define TEMPERATURE_READ_FAILED  1
```
Do we need to define such label? Wouldn't it more adequate to return the value returned by sensor API functions (or even `-EIO`)?

gamnes replied:
No, we don't. I will remove it and use the return value of the sensor API.

anangl commented on samples/boards/nrf52/sensor_temperature/src/main.c:
```diff
@@ -0,0 +1,97 @@
+		}
+		/* Sensor reading will be integer val1, and fractional val2.
+		 * The value returned from the nRF will be in the form off
+		 * val1 = 22, val2 = 250000, which is 22.5 degC.
```
This comment is a bit misleading. In these sample values either the `val2` should be `500000` or the tempearture `22.25`.
But perhaps it would be better to provide instead a comment to the line where `val2` is divided by `10000` (that this is the fractional part of the readout value, in one-millionth parts).

gamnes replied:
Correct, typo from my side. Will change the val2 to be 50....
I will look at moving the comment somewhere else. I basically wrote this comment for my own understanding whilst writing the code.

anangl commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -133,6 +134,18 @@
+		 */
+		temp: temp@4000C000 {
+			compatible = "nordic,nrf-temp";
+			reg = <0x4000C000 0x1000>;
```
The addresses should be in lowercase. See https://github.com/zephyrproject-rtos/zephyr/commit/523932ccb0a6dadbef0b273150ee3e80b906e915.

gamnes replied:
(Y) Just so used to writing hex symbols as uppercase. Will change it.

anangl commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -133,6 +134,18 @@
+			reg = <0x4000C000 0x1000>;
+			interrupts = <12 1>;
+			status = "ok";
+		};
```
Shouldn't we provide a label property for the node, like for all others?

gamnes replied:
Currently, the label property for this node is not used for anything - so i did not see any reasons to add it. This is one of the questions I have been wondering about, should a driver label/name be defined in Kconfig - or should it be define in DTS? 

The driver name for this TEMP driver is currently CONFIG_TEMP_NRF5_NAME, ref: https://github.com/zephyrproject-rtos/zephyr/pull/12034/files#diff-4adad1dc7c1f3fd67902b233f21c7ef2R128. For me, it makes sense that a SW drivers name is defined in Kconfig because that is where a user will enable the driver - and thus at the time they enable it, they need to know the name of the driver so that they can use it in their code. If the name is defined in DTS, he will have to search through the DTS file to find the name after enabling it through Kconfig - that seems like more of a hassle? 

I can add a label just to align it with the other nodes, but it wont be used for anything.

gamnes replied:
Added label property for TEMP node.

gamnes commented on drivers/sensor/nrf5/Kconfig:
```diff
@@ -9,6 +9,7 @@
 menuconfig TEMP_NRF5
 	bool "nRF5 Temperature Sensor"
 	depends on SOC_FAMILY_NRF
+	depends on SENSOR
```
The reason I did this was because during my testing, I thought I only needed to set CONFIG_TEMP_NRF5=y in my prj.conf file to enable the TEMP_NRF5 sensor; as there is no reference from this temp_nrf5 Kconfig file to the overall SENSOR Kconfig file. But yes, I agree (Y)

gamnes commented on drivers/sensor/nrf5/temp_nrf5.c:
```diff
@@ -116,11 +109,26 @@ static int temp_nrf5_init(struct device *dev)
+static struct temp_nrf5_data temp_nrf5_driver;
+
+DEVICE_AND_API_INIT(temp_nrf5,
+					CONFIG_TEMP_NRF5_NAME,
```
Yupp. Fixed.

gamnes commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -133,6 +134,18 @@
+		 * .. Interrupt priority:
+		 * .. .. Range: 0-1, Default: 1
+		 */
+		temp: temp@4000C000 {
```
But this line stays UPPER case C? Or lower case as well?

anangl replied:
I believe both occurrences of the address should be corrected to lowercase. As it was done in the commit I mentioned.

gamnes replied:
Ok, then we should also change the flash-controller@4001E000 of this file as well i suppose? 
I don't understand why we don't accept both uppercase and lowercase (assuming this is a tool "issue"), but I will make the change for this temp node.

gamnes replied:
Changed TEMP node addresses to lower case.

gamnes commented on samples/boards/nrf52/sensor_temperature/src/main.c:
```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright (c) 2016 Intel Corporation
```
Your are correct (Y) fixed.

gamnes commented on drivers/sensor/nrf5/temp_nrf5.c:
```diff
@@ -16,6 +16,9 @@ LOG_MODULE_REGISTER(TEMPNRF5);
 
 
+static NRF_TEMP_Type *const nrf_temp_addr =
+		(NRF_TEMP_Type *)DT_NORDIC_NRF_TEMP_TEMP_BASE_ADDRESS;
```
Agreed. I thought it was decided that this was the way to go based on the nrfx_uart.c driver implementation. I will revert to using the NRF_TEMP. Then this can be revisited whenever the issue you referenced is resolved.

gamnes replied:
But i will be removing the volatile *temp pointer from the functions as i dont see the need to create a pointer to the struct that is used. Agree? So the temp_nrf5_sample_fetch() function will have "volatile NRF_TEMP_Type *temp = NRF_TEMP;" removed, and "temp->TASKS_START = 1;" will be "NRF_TEMP->TASKS_START = 1;" instead.

anangl replied:
> But i will be removing the volatile *temp pointer from the functions as i dont see the need to create a pointer to the struct that is used. Agree?

Yes. I have no idea why this pointer was introduced. I don't see a need for it, too. It only obscures the code IMHO.

gamnes commented on drivers/sensor/nrf5/temp_nrf5.c:
```diff
@@ -16,6 +16,9 @@ LOG_MODULE_REGISTER(TEMPNRF5);
 #include "nrf_common.h"
```
Removed.

gamnes commented on samples/boards/nrf52/sensor_temperature/src/main.c:
```diff
@@ -0,0 +1,97 @@
+#define LED1		DT_GPIO_LEDS_LED1_GPIO_PIN
+#define LED2		DT_GPIO_LEDS_LED2_GPIO_PIN
+#define LED3		DT_GPIO_LEDS_LED3_GPIO_PIN
+int led_write_values[4] = {0};
```
Removed (Y)

gamnes commented on samples/boards/nrf52/sensor_temperature/src/main.c:
```diff
@@ -0,0 +1,97 @@
+	gpio_pin_configure(dev, LED0, GPIO_DIR_OUT);
+	gpio_pin_configure(dev, LED1, GPIO_DIR_OUT);
+	gpio_pin_configure(dev, LED2, GPIO_DIR_OUT);
+	gpio_pin_configure(dev, LED3, GPIO_DIR_OUT);
```
So, should i be getting/creating LED Driver binding for each and every LED? 
Will this LED driver binding be "attached" to the same driver - since they are in fact connected to the same GPIO port? How do you handle this if you have more, like 32, GPIOs with LEDs? Will every single one get its one driver binding?

gamnes replied:
Using DT_GPIO_P0_DEV_NAME from ...nordic_nrf\nrf52\dts_fixup.h. Not sure if this makes a difference as I am still assuming (/know) that all GPIOs are on the same GPIO P0 port.

anangl replied:
No, it does not make much difference. My point is that if you decide to get information about LEDs from DTS, it does not look good when you then assume that all the LEDs are on the same port. You should rather use `DT_GPIO_LEDS_LEDx_GPIO_CONTROLLER` for each LED. The same applies to their active state. It should also be taken from DTS, and currently the code assumes that `0` is always the active state.
On the other hand, using all this information from DTS will blow the code of this fairly simple application. So I'm not convinced it is worth it. Maybe this controlling of the LEDS could be dropped and just printing the temperature would be enough. It does not seem to be easy to see the temperature crossing the 5 or 10 degrees threshold on your desk, so maybe this LED signaling will not be that useful.

gamnes replied:
Example removed.

ioannisg reviewed (commented):
I have a question regarding the .dtsi changes; would it make more sense to declare the status as "disabled" and set it to "ok" in the board dts?

gamnes commented:
> I have a question regarding the .dtsi changes; would it make more sense to declare the status as "disabled" and set it to "ok" in the board dts?

Yes, that would make more sense. But then does it make even more sense to enable it in the application folder through DTS overlay - since not every application using a board will be using the TEMP sensor?

gamnes committed:
drivers: nrf5: temp: Cleanup temp_nrf5.c source

1. Remove nrf_common.h include
2. Remove unnecessary NRF_*Type defines, using
   CMSIS NRF_TEMP define directly instead.
3. Align driver code by including DEVICE_DECLARE,
   and moving DEVICE_AND_API_INIT() to the bottom.

Signed-off-by: Gaute Gamnes <gaute.gamnes@nordicsemi.no>

gamnes committed:
dts: nrf: Temp device node added to nRF5 devices with yaml binding

1. Temp device node added to all nRF5 DTSI files.
2. Added yaml binding for Nordic Temp node.
3. Set codeowner of nordic dts bindings to @anangl

Signed-off-by: Gaute Gamnes <gaute.gamnes@nordicsemi.no>

gamnes committed:
drivers: nrf5: temp: Migrate Kconfig priority option to DT

1. Kconfig options priority setting removed.
2. Modified temp_nrf5.c driver to use DT
   defines instead of Kconfig.

Signed-off-by: Gaute Gamnes <gaute.gamnes@nordicsemi.no>

anangl commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -140,6 +141,14 @@
+			reg = <0x4000c000 0x1000>;
+			interrupts = <12 1>;
+			status = "ok";
+			label = "TEMP";
```
"TEMP_0", to stay consistent with the default value of `CONFIG_TEMP_NRF5_NAME`.

gamnes replied:
Changed to TEMP_0.

anangl commented on dts/bindings/sensor/nordic,nrf-temp.yaml:
```diff
@@ -0,0 +1,32 @@
+      type: array
+      description: mmio register space
+      generation: define
+      category: required
```
This is just a detail but since this is a new file, I think it would be good to make it look nicer and place this key right after "type", to keep consistency with other properties. Also the empty line before `...` seems to be unneeded.

gamnes replied:
Agreed. Moved up. 
Some files have an empty line before the ..., some do not, and some don't have these dots either (what's their purpose?). But i removed the empty line for this PR.

anangl replied:
AFAICT these periods are the optional end of a yaml document within a stream. We probably don't need them at all (as well as the opening `---`) in DTS bindings. But they are present in most of these files. I'm not an expert in this field.

gamnes commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -47,6 +47,7 @@
 		gpiote-0 = &gpiote;
 		wdt-0 = &wdt;
 		qdec-0 = &qdec;
+		temp = &temp;
```
Removed - defines in temp_nrf5.c changed to match the new ones that were automatically generated.
